### PR TITLE
Make RepositoryData Less Memory Heavy

### PR DESCRIPTION
**Pull request by original-brownbear**
We don't really need `LinkedHashSet` here. We can assume that all the
entries are unique and just use a list and use the list utilities to
create the cheapest possible version of the list.
Also, this fixes a bug in `addSnapshot` which would mutate the existing
linked hash set on the current instance (fortunately this never caused a real world bug)
and brings the collection in line with the java docs on its getter that claim immutability.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java [hunk 1]
```diff
@@ -213,9 +214,16 @@ public RepositoryData addSnapshot(final SnapshotId snapshotId,
         newSnapshotStates.put(snapshotId.getUUID(), snapshotState);
         Map<String, Version> newSnapshotVersions = new HashMap<>(snapshotVersions);
         newSnapshotVersions.put(snapshotId.getUUID(), version);
-        Map<IndexId, Set<SnapshotId>> allIndexSnapshots = new HashMap<>(indexSnapshots);
+        Map<IndexId, List<SnapshotId>> allIndexSnapshots = new HashMap<>(indexSnapshots);
         for (final IndexId indexId : shardGenerations.indices()) {
-            allIndexSnapshots.computeIfAbsent(indexId, k -> new LinkedHashSet<>()).add(snapshotId);
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:218]: This was broken, we were mutating the existing `LinkedHashSet`

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java [hunk 2]
```diff
@@ -253,23 +261,24 @@ public RepositoryData removeSnapshot(final SnapshotId snapshotId, final ShardGen
...
         newSnapshotVersions.remove(snapshotId.getUUID());
-        Map<IndexId, Set<SnapshotId>> indexSnapshots = new HashMap<>();
+        Map<IndexId, List<SnapshotId>> indexSnapshots = new HashMap<>();
         for (final IndexId indexId : indices.values()) {
-            Set<SnapshotId> set;
-            Set<SnapshotId> snapshotIds = this.indexSnapshots.get(indexId);
+            List<SnapshotId> remaining;
+            List<SnapshotId> snapshotIds = this.indexSnapshots.get(indexId);
             assert snapshotIds != null;
             if (snapshotIds.contains(snapshotId)) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:269]: Not great that we're quadratic here now (for the nested loop), but I don't think it really matters much relative to the significant space+GC savings.

**Comment by original-brownbear** [comment]
Found this while examining a heap dump for a cluster running into #55153 .
For a few thousand indices and ~1k snapshots the heap was full of uncollected `java.util.LinkedHashMap.Entry` (the sets/maps themselves were mostly already collected but collecting the entries is apparently trickier for the JVM) under load on the snapshot status APIs from the somewhat long lived `RepositoryData` instances. Also the outright overhead is massive for `LinkedHashSet` just on the face of it with at least `44` bytes per element (say the cluster contains 100 live indices at a time and we have 100 snapshots this comes out to half a MB in overhead per instance easily).

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java [hunk 3]
```diff
@@ -213,9 +214,16 @@ public RepositoryData addSnapshot(final SnapshotId snapshotId,
...
+        Map<IndexId, List<SnapshotId>> allIndexSnapshots = new HashMap<>(indexSnapshots);
         for (final IndexId indexId : shardGenerations.indices()) {
-            allIndexSnapshots.computeIfAbsent(indexId, k -> new LinkedHashSet<>()).add(snapshotId);
+            final List<SnapshotId> snapshotIds = allIndexSnapshots.get(indexId);
+            if (snapshotIds == null) {
+                allIndexSnapshots.put(indexId, List.of(snapshotId));
+            } else {
+                final List<SnapshotId> copy = new ArrayList<>(snapshotIds);
+                copy.add(snapshotId);
+                allIndexSnapshots.put(indexId, List.copyOf(copy));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:225]: why create a copy of the copy?

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java [hunk 4]
```diff
@@ -253,23 +261,25 @@ public RepositoryData removeSnapshot(final SnapshotId snapshotId, final ShardGen
...
                 if (snapshotIds.size() == 1) {
                     // removing the snapshot will mean no more snapshots
                     // have this index, so just skip over it
                     continue;
                 }
-                set = new LinkedHashSet<>(snapshotIds);
-                set.remove(snapshotId);
+                remaining = new ArrayList<>(snapshotIds);
+                remaining.remove(listIndex);
+                remaining = List.copyOf(remaining);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:278]: same thing here, copy of copy

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java [hunk 5]
```diff
@@ -513,7 +523,7 @@ public static RepositoryData snapshotsFromXContent(final XContentParser parser,
                             }
                         }
                         assert indexId != null;
-                        indexSnapshots.put(indexId, snapshotIds);
+                        indexSnapshots.put(indexId, List.copyOf(snapshotIds));
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:526]: copy of copy

**Review by ywelsch — COMMENTED** [review]
Change looks ok, but there is too much unnecessary list copying going on.

File: server/src/main/java/org/elasticsearch/repositories/RepositoryData.java (hunk 3 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:225]: These `RepositoryData` instances live for quite a while, so I figured the cost of doing another copy is worth the lower storage overhead + shorter path to the GC root compared to wrapping with `Collections.unmodifiableList`? I could technically make this more efficient by copying to a `SnapshotId[]` and then just wrapping that array but I figured this wasn't that much slower and nicer to read.

Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/repositories/RepositoryData.java:225]: I looked into how `List.copyOf` is implemented, and lo and behold, it copies the elements twice (first calls `Collection.toArray()`, and then creates another copy of that temporary array in `List.of` (using manual for loop, FFS).
This means that the list is copied three times here, plus the resize of the `ArrayList` when calling `copy.add(snapshotId);`, leading to another full copy ....
High-level languages ftw.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/RepositoryData.java]: :) you win => I pushed https://github.com/elastic/elasticsearch/pull/55293/commits/8319e2116d138668e38c3a72b709a8889868088a , probably not worth the hassle to go further than this then.

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Yannick!

**Pull request merged**